src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs

```diff
@@ -0,0 +1,85 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+/// <summary>
+/// Extension methods for annotating OpenAPI descriptions on an <see cref="Endpoint" />.
+/// </summary>
+public static class OpenApiRouteHandlerBuilderExtensions
+{
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+
+    public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder)
+    {
+        builder.Add(endpointBuilder =>
+        {
+            if (endpointBuilder is RouteEndpointBuilder routeEndpointBuilder)
+            {
+                var openApiOperation = GetOperationForEndpoint(routeEndpointBuilder);
+                if (openApiOperation != null)
+                {
+                    routeEndpointBuilder.Metadata.Add(openApiOperation);
+                }
+            };
+        });
+        return builder;
+
+    }
+
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint and modifies it with the given <paramref name="configureOperation"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <param name="configureOperation">An <see cref="Action"/> that mutates an OpenAPI annotation.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+    public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder, Func<OpenApiOperation, OpenApiOperation> configureOperation)
```

Comment:
Do we have any tests that call the WithOpenApi overloads? The ones I see all call into OpenApiGenerator directly.

Reply:
Added some test coverage here.